File: 프로그래머스/2/388352. 비밀 코드 해독/비밀 코드 해독.java
import java.util.*;

class Solution {
    private int n;
    private int[][] q;
    private int[] ans;
    private int answer;
    
    private void comb(int start, List<Integer> current) {
        if (current.size() == 5){
            if (validate(current)) answer++;
            return;
        }
        
        for (int i=start; i<=n; i++) {
            current.add(i);
            comb(i+1, current);
            current.remove(current.size()-1);
        }
    }
    
    private boolean validate(List<Integer> combination) {
        Set<Integer> nums = new HashSet<>(combination);
        
        for (int i = 0; i < q.length; i++) {
            int count = 0;
            for (int num : q[i]) if (nums.contains(num)) count++;
            if (count != ans[i]) return false;
        }
        return true;
    }
    
    public int solution(int n, int[][] q, int[] ans) {
        this.n = n;
        this.q = q;
        this.ans = ans;
        answer = 0;
        comb(1, new ArrayList<Integer>());
        return answer;
    }
}

File: 프로그래머스/2/388352. 비밀 코드 해독/비밀 코드 해독.kt
class Solution {
    private var n: Int = 0
    private lateinit var q: Array<IntArray>
    private lateinit var ans: IntArray
    private var answer = 0
    
    private fun comb(start: Int, current: MutableList<Int>) {
        if (current.size == 5) {
            if (validate(current)) answer++
            return
        }
        for (i in start..n) {
            current.add(i)
            comb(i+1, current)
            current.removeAt(current.size-1)
        }
    }
    
    private fun validate(comb: List<Int>): Boolean {
        val set = comb.toHashSet()
        
        for (i in q.indices) {
            var cnt = 0
            for (num in q[i]) if (num in set) cnt++
            if (cnt != ans[i]) return false
        }
        return true
    }
    
    fun solution(n: Int, q: Array<IntArray>, ans: IntArray): Int {
        this.n = n
        this.q = q
        this.ans = ans
        answer = 0
        comb(1, mutableListOf())
        return answer
    }
}
